Adding the MapNotifier class

mboyamike opened the pull request:
A MapNotifier class with its tests. It behaves like a map but notifies all its listeners whenever a change is made.
I am yet to implement the don't notify if equal and the custom equality. I'll first try to understand how it had been implemented in the ListNotifier class.

mboyamike committed:
Creating the MapNotifier class

This behaves in many ways like the normal Map, but notifies listeners whenever a change occurs.

mboyamike committed:
Tests for the MapNotifier class

The first batch of tests. More tests coming up

mboyamike committed:
Added export for map_notifier

escamoteur commented on lib/src/map_notifier.dart:
```diff
@@ -0,0 +1,153 @@
+  @override
+  void operator []=(K key, V value) {
+    startTransAction();
+    super[key] = value;
```
Hi, 
why do you wrap this and the following functions in ´start/stopTransactions´?
As long as only one element of the Map is changed, a call to _notify should be enough.

escamoteur commented on lib/src/map_notifier.dart:
```diff
@@ -0,0 +1,153 @@
+    endTransAction();
+  }
+
+  @override
```
As we are extending ´DelegatingMap´ we don't have to overwrite methods that we don't change in behabviour. we can just omit this overwride.

escamoteur commented on lib/src/map_notifier.dart:
```diff
@@ -0,0 +1,153 @@
+      super.map(transform);
+
+  @override
+  V putIfAbsent(K key, V Function() ifAbsent) {
```
that's an interesting one. actually we only want to call _notify if something really got changed. 
Have you tried if this method would work correctly if we just removed it and let the implementation of the super class implement it? I don't know it the implmentation of the super class uses the ´[]´operator so that automatically _notify should be called because we overrode the setter ´[]`above. 
Please give it a try. 
if not, we only have to call _notify if the ífAbsent´parameter is called.

escamoteur replied:
did you test this?

mboyamike replied:
In the end it was extra work. Using the implementation of the super class was better.

escamoteur commented on lib/src/map_notifier.dart:
```diff
@@ -0,0 +1,153 @@
+  }
+
+  @override
+  V remove(Object key) {
```
as with the other methods no wrapping with ´start/stopTransaction´ we only call notify if the map was modified , which means only if the object was in the collection. So you have to check ´v´ if it's not null before you call _notify.

escamoteur commented on lib/src/map_notifier.dart:
```diff
@@ -0,0 +1,153 @@
+  }
+
+  @override
+  void removeWhere(bool Function(K, V) test) {
```
another interesting one. we only want to notify if we change the map. So store the length of the map before you call the super.removeWhere and compare it to the length afterwards.

mboyamike replied:
This new update does that. Thanks

escamoteur commented on lib/src/map_notifier.dart:
```diff
@@ -0,0 +1,153 @@
+  }
+  
+  @override
+  Iterable<V> get values => super.values;
```
Unnecessary overrides

escamoteur commented on lib/src/map_notifier.dart:
```diff
@@ -0,0 +1,153 @@
+  String toString() => super.toString();
+
+  @override
+  V update(K key, V Function(V) update, {V Function() ifAbsent}) {
```
here either  ´udate´or ´ifabsent´is executed. so instead of directly passing them to super we pass
´´´Dart
(){
 final v = upDate();
 _notify();
return v;
´´´
same for ´ifAbsent´

escamoteur commented on lib/src/map_notifier.dart:
```diff
@@ -0,0 +1,153 @@
+  }
+
+  @override
+  void updateAll(V Function(K, V) update) {
```
Here we can't easily say if the collection has changed or not. In this case I would all ´notify()´always after the call to super

escamoteur replied:
Same as above does this trigger multiple notifications with out the Transactions?

mboyamike replied:
After testing, it doesn't.
I think it was a case of making assumptions about how the code works.

escamoteur commented on test/map_notifier_test.dart:
```diff
@@ -0,0 +1,40 @@
+      buildListener();
+      mapNotifier.clear();
+      expect(result, {});
+    });
```
because we added new behaviour to the methods acording to my comment, we will need some more Tests. One test for every overridden function. Also think what could go wrong when calling such a function and write a test for it.

escamoteur reviewed (changes requested):
Hi,

many thanks for the contribution. I added some comments on what to change.
If something isn't clear, just ask.
thomas

mboyamike committed:
Remove unnecessary overrides and transactions

Replacing the unnecessary transactions and replacing them with a call to notify. Also removed the overrides from the super class that did not alter behavious.
Yet to implement the custom equality and the notify if equal part. Tell me what you think so far.

mboyamike committed:
Adding tests

Added a few tests for the map notifier

escamoteur reviewed (approved):
Looks good

escamoteur commented on lib/src/map_notifier.dart:
```diff
@@ -48,106 +48,69 @@ class MapNotifier<K, V> extends DelegatingMap<K, V> with ChangeNotifier
     V v = super.remove(key);
-    endTransAction();
+
+    if(this.value.length != initialLength)
```
Please always put {} around the line an if is for.

mboyamike replied:
Thanks for the tip! 
Experience came to teach me the importance of that :)


I'm resolving the old conversations cause I made several changes

escamoteur commented on lib/src/map_notifier.dart:
```diff
@@ -48,106 +48,69 @@ class MapNotifier<K, V> extends DelegatingMap<K, V> with ChangeNotifier
-    endTransAction();
+
+    /// only notify if an element was removed
+    if(this.value.length == initialLength)
```
your comment is right but the test is wrong. shouldn't this be != ?
And please add {}

escamoteur commented on lib/src/map_notifier.dart:
```diff
@@ -48,106 +48,69 @@ class MapNotifier<K, V> extends DelegatingMap<K, V> with ChangeNotifier
-
   @override
   V update(K key, V Function(V) update, {V Function() ifAbsent}) {
-    startTransAction();
```
Have you tested if this works without transactions if `update` changes something?

escamoteur reviewed (commented):
Would like some replys to the comments

escamoteur commented:
hi @mboyamike do you still want to finish this PR?

mboyamike commented:
Hi @escamoteur 

Sorry, I got a bit held up with work and completely forgot about it. 
Yes, I'd like to finish this PR. I'll submit changes by end of week.

escamoteur commented:
you will have probably make some changes because I migrated to null safety today
Looking forward to your contribution!

mboyamike commented:
No worries, thanks

mboyamike committed:
Merge branch 'escamoteur:master' into master

mboyamike committed:
map notifier full tests and upgrade to null safety

mboyamike commented:
Hey @escamoteur 
Sorry for my long departure. Finally got back to it.
I've migrated everything to null safety. Added the MapNotifier. By using a TDD driven approach I was able to ensure each line is tested to be required. I still have a bit of comments to write as part of documentation, and also think of a way to implement the custom equality in them.
A question on custom equality, will we only want to compare values or should the keys also be considered?

escamoteur commented:
Hey, cool that you are back. Only values, keays don't make sense because that uses the hashCode function that people can overload on their type
Am 30. März 2022, 07:14 +0200 schrieb Mboya Michael ***@***.***>:
> @mboyamike commented on this pull request.
> In lib/src/map_notifier.dart:
> > +  @override
> +  Iterable<V> get values => super.values;
> +
> +  @override
> +  String toString() => super.toString();
> +
> +  @override
> +  V update(K key, V Function(V) update, {V Function() ifAbsent}) {
> +    startTransAction();
> +    V v = super.update(key, update, ifAbsent: ifAbsent);
> +    endTransAction();
> +    return v;
> +  }
> +
> +  @override
> +  void updateAll(V Function(K, V) update) {
> After testing, it doesn't.
> I think it was a case of making assumptions about how the code works.
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you were mentioned.Message ID: ***@***.***>

mboyamike commented:
Ok great, thanks. I'll implement that

mboyamike committed:
Add Custom Equality

mboyamike committed:
simplify the should notify method.

mboyamike committed:
Add documentation comments

mboyamike commented:
Added documentation and equality

escamoteur commented on lib/src/map_notifier.dart:
```diff
@@ -0,0 +1,181 @@
+
+    if (_shouldNotify(initialMapValue)) {
+      notifyListeners();
+    }
```
Sorry, that it took me so long! Looks really good. I just wonder as be repeat this 

```
    if (_shouldNotify(initialMapValue)) {
      notifyListeners();
    }
```
everywhere why not move that to the `_shouldNotify` function and maybe rename it to `_checkNotificationNeeded` for instance?

If I don't get back to you in time, please ping me because I easily loos track of this when I have a not so good time.
Cheers
Thomas

mboyamike replied:
No worries. Thank you.
Yeah, let me make that change. It's a cleaner approach.
Take care!
Mboya

escamoteur reviewed (changes requested):
If I don't get back to you in time, please ping me because I easily loose track of this when I have a not so good time.
Cheers
Thomas

mboyamike committed:
introduce _notifyIfNeeded to reduce duplication

mboyamike commented:
Hey @escamoteur 
When you get some time, could you check the current change. Added a _notifyIfNeeded method

escamoteur commented:
Hi, I'm really sorry that I completely lost this out or my mind, will check soon
